### ISSUE TITLE
CMake install headers & Meson intermediate install

### ISSUE THREAD

**Issue by magnusmarc** [mesonbuild/meson#10298]
**Describe the bug**
In reference to #2546.

There is a problem where meson does not install headers during intermediary build steps. If you have a library "library_test" with a folder "src/" and all the includes & src files in that directory-- you will have to access those files like
`#include <yourfile.h>`
regardless of your install directory. A good example is wanting them installed under "library_test/yourfile.h"
`#include <library_test/yourfile.h>` will not work.

The workaround does work, but it is not clean and has limitations.

**The main problem** I am seeing is when CMake tries to do intermediary build steps. Cmake normally has INSTALL_HEADERS that are done during the build which can be used by your program to link correctly. However, Meson does not have an equivalent feature, but an equivalent workaround. (which does not generate with the cmake module)

```
SET(hs_HEADERS
    src/hs.h
    src/hs_common.h
    src/hs_compile.h
    src/hs_runtime.h
)
install(FILES ${hs_HEADERS} DESTINATION "${CMAKE_INSTALL_INCLUDEDIR}/hs")
```

In short, calling a Cmake library that relies on transferring headers to the correct location is not done and forces the user to change their includes to however the file is positioned in the source directory.

**To Reproduce**
https://github.com/intel/hyperscan 
```
cmake = import('cmake')
yourdep = cmake.subproject('hyperscan', options: hyperscan_proj_opts, required: true).dependency('hs')
```

**Expected behavior**
`#include <hs/hs.h>` is the correct include but it is not found. Instead `#include <hs.h>` has to be used instead.

**system parameters**
* Is this a [cross build](https://mesonbuild.com/Cross-compilation.html) or just a plain native build (for the same computer)?
Native.
* what operating system (e.g. MacOS Catalina, Windows 10, CentOS 8.0, Ubuntu 18.04, etc.)
Ubuntu 18.04
* what Python version are you using e.g. 3.8.0
Python 3.6.9
* what `meson --version`
0.61.3
* what `ninja --version` if it's a Ninja build
1.8.2

Thanks for the help! Let me know if you need more info.

**Comment by eli-schwartz** [comment]
Really, having headers in a different source vs. installed layout isn't a great idea.

**Comment by magnusmarc** [comment]
@eli-schwartz, some projects go with this design. Only choice is to redesign the way they build & their source code or work around it. The install directory is the way that the program is actually run, so either we shouldn't give an option for install directory configuration or workaround it.

**I agree that the folder should be the same for the include**. It is harder to maintain visibility on includes when it is in a install vs an actual source directory. Any projects I work on have include/<subdir>/<headers>. Then the <subdir> is the folder that is used for includes and works fine.

Many non meson libraries will follow the old design route which is having many source and include files in the same directory-- then install the layout differently. The old build systems don't have as many checks in place and everything was built differently. Rather than having meson manage a group of subprojects, it would just be many Makefiles being called with a install directory being passed around for later linking.

**This is why I prefer Meson :)**. But it still makes it difficult to work with non-Meson systems which do not follow the same approach.

**Comment by eli-schwartz** [comment]
Right, and this is assigned to our cmake maintainer in the hope that some solution can be figured out. :)

(It's always worth trying to beg upstream for some sanity, but granted that may not end up succeeding.)

One possible solution might be to do the equivalent of `configure_file(..., copy: true)` as per the "equivalent workaround" under the hood for any files cmake marks for installation, but I know basically nothing about how the cmake module actually works so I couldn't say for sure.

**Comment by mensinda** [comment]
## Regarding adding support for this

There are currently multiple problems with trying to "fix" this on the Meson side (ignoring the whole discussion on whether this should be fixed upstream):

- The current Meson CMake backend does not really handle installing subprojects really well in general. For instance, Meson ignores the `install(FILES ${hs_HEADERS} DESTINATION "${CMAKE_INSTALL_INCLUDEDIR}/hs")` line currently.

  To get this to work, we would have to add a new parser for the CMake `install` command. Then we have to correctly and reliably detect `${CMAKE_INSTALL_INCLUDEDIR}` (which is not trivial because all variables get expanded and Meson will only ever see the full path). Next, we have to copy the file at configure time and add the destination directory as a global include directory.

  Each step is doable, but it adds up with time. Additionally, it does come with a significant maintenance burden and *will* slow down the configuration time of CMake subprojects due to the added copying overhead.
- In a Nutshell, the Meson CMake backend works like this:
    1. Call CMake (with some special parameters) and let it configure the project
    2. Parse and interprete the metadata the CMake produces
    3. Build a Meson AST (which is also stored at `build/subprojects/hyperscan/meson.build` for easy debugging)
    4. Pretend that we now have a Meson project and do the normal subproject logic

  This means that CMake subprojects support everything that Meson supports. Nothing more and nothing less. In this case, this is a serious limitation because we could only reasonably generate the `configure_file(..., copy: true)` stuff. However, this will completely break down for generated headers.

---

## Workaround

As a workaround, you could add a dummy include directory with dummy header files:

```c
// dummy_include/hs/hs.h
#pragma once
#include <hs.h>
```

```meson
if use_subproject:
  cmake = import('cmake')
  yourdep = cmake.subproject('hyperscan', required: true).dependency('hs')
  global_inc = include_directories('dummy_include', is_system: true)
endif
```

I have not tested this yet, but it should work (even though it is ugly...). If this (or something like it) works, then I don't really see a good argument for trying to fix this on a Meson level (especially because of the performance implications).

**Comment by magnusmarc** [comment]
@mensinda , that would work. For future Cmake projects I suppose this could be done. The best workaround in my opinion is to have a cmake conversion script that will properly convert the entire project to meson and make manual modifications to fix issues like this and upload the files to the WrapDB system.

It would be great if the Cmake projects generated Meson files that did not have hard coded paths (I think they do have hard coded paths right now).

**Comment by mensinda** [comment]
@magnusmarc You could try playing around with the install options in [the CMake subproject configuration object](https://mesonbuild.com/CMake-module.html#cmake-options-object). But if that also doesn't work, then this will probably still be a wontfix...

**Comment by HenningScheufler** [comment]
I have two cmake-libraries in my project: hdf5 and cgns. Hdf5 does not compile with the meson's cmake-module and I found this not really beautiful workaround:
```python
r = run_command(['cmake',
                '-G Ninja',
                '-S .',
                '-B build',
                '-DCMAKE_INSTALL_PREFIX=' + get_option('prefix')], check: true)

r = run_command(['cmake',
                '--build',
                'build'], check: true)

r = run_command(['cmake',
                '--install',
                'build'], check: true)
```

cgns does compile with the meson cmake module but does **not** install the header. So, I tend to duplicate the above approach. Is there a more elegant way?

**Comment by haata** [comment]
> 
> `hs_includes = files('subprojects/hyperscan-5.4.0/src/hs.h', ...)`
> 

As a note, at least with meson v1.1.1 this doesn't seem possible from the root project.

```
../../../meson.build:148:18: ERROR: Sandbox violation: Tried to grab file zookeeper.h from a nested subproject.
```

I'm not sure how to do it, but one way to deal with this problem would be to allow a list of files to be provided to the cmake subproject so that the top-level project could do the install itself.

For example
```meson
zookeeper_options = cmake.subproject_options()
zookeeper_options.add_install_files(['zookeeper-client/zookeeper-client-c/include/zookeeper.h'], prefix: 'include')
zookeeper = cmake.subproject('zookeeper', options: zookeeper_options)
```

Or something like that.

In my particular case I already have a wrapper script around meson (to handle some more complicated subproject wrap situations; and some extra warnings/errors) so I'll just have those scripts pull out the necessary headers.
